src/docfx/restore/RestoreGit.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class RestoreGit
+    {
+        private const int MaxWorkTreeCount = 5;
+
+        public static string GetRestoreWorkTreeDir(string restoreDir, string workTreeHead)
+            => PathUtility.NormalizeFile(Path.Combine(restoreDir, workTreeHead));
+
+        public static string GetRestoreRootDir(string url)
+            => Docs.Build.Restore.GetRestoreRootDir(url, AppData.GitRestoreDir);
+
+        public static async Task<IEnumerable<(string href, string workTreeHead)>> Restore(string docsetPath, Config config, Func<string, Task> restoreChild)
+        {
+            var workTreeMappings = new ConcurrentBag<(string href, string workTreeHead)>();
+            var restoreItems = config.Dependencies.Values.GroupBy(d => GetRestoreRootDir(d)).Select(g => (g.Key, g.Distinct().ToList()));
+
+            // process git restore items
+            await ParallelUtility.ForEach(
+               restoreItems,
+               async restoreItem =>
+               {
+                   var (restoreDir, hrefs) = restoreItem;
+                   var workTreeHeads = await RestoreDependentRepo(restoreDir, hrefs);
+                   foreach (var workTreeHead in workTreeHeads)
+                   {
+                       workTreeMappings.Add(workTreeHead);
+                   }
+               },
+               progress: Progress.Update);
+
+            return workTreeMappings;
+
+            async Task<List<(string href, string head)>> RestoreDependentRepo(string restoreDir, List<string> hrefs)
+            {
+                var workTreeHeads = await GetWorkTrees(docsetPath, restoreDir, hrefs);
+
+                foreach (var (_, workTreeHead) in workTreeHeads)
+                {
+                    var childDir = GetRestoreWorkTreeDir(restoreDir, workTreeHead);
+                    await restoreChild(childDir);
+                }
+
+                return workTreeHeads;
+            }
+        }
+
+        // Restore dependent repo to local and create work tree
+        private static async Task<List<(string href, string head)>> GetWorkTrees(string docsetPath, string restoreDir, List<string> hrefs)
```

Comment:
>GetWorkTrees [](start = 68, length = 12)

It seems hard to read to put local functions in the middle of a function... Move to the end? 

I also think it's hard to read for too many local functions in a method...